### PR TITLE
HHH-13831 Replaced listener not called when event is fired

### PR DESCRIPTION
**Pull request by DavideD**
When replacing a listener the array in `EventListenerGroupImpl`  is not refereshed.

Is this a bug or am I missing something?

I've created a JIRA: https://hibernate.atlassian.net/browse/HHH-13831

Thanks

### CODE REVIEW

**Comment by jwgmeligmeyling** [comment]
Probably an oversight / it just wasn't necessary before. As it is in the internal package, I wouldn't consider this a bug. Whats the use case? The unit test seems very specific.

Besides that, I am all for integrating this if it solves any issue for you.

**Comment by DavideD** [comment]
The replaced listener is not called if the event is fired using `EventListenerGroup#fireEventOnEachListener`
We didn't notice before because most of the time `EventListenerGroup#listeners()` is called.

This happend because the listener is replaced in internal field `listenersAsList` but it stays out of sync with the other array field `listeners`. The patch mark the `listeners` field for refresh when we replace one of the listeners.

File: hibernate-core/src/test/java/org/hibernate/event/service/internal/EventListenerReplacementStrategyTest.java [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+		}
+	}
+
+	private static class ReplaceOriginalStrategy implements DuplicationStrategy {
+
+		static final ReplaceOriginalStrategy INSTANCE = new ReplaceOriginalStrategy();
+
+		@Override
+		public boolean areMatch(Object listener, Object original) {
+			return original.getClass().equals( OriginalListener.class )
```
Comment (sebersole) [inline hibernate-core/src/test/java/org/hibernate/event/service/internal/EventListenerReplacementStrategyTest.java]: I'm a little nervous about this.  Why not just check for each being an `instanceof ClearEventListener`?  The intention of your test is clearly that there be only one.

I'd also consider asserting `org.hibernate.event.service.spi.EventListenerGroup#listeners` to make sure that the listener actually got replaced/appended.  Coupled with the point above, its possible that your code itself is not actually appending/replacing the listener

File: hibernate-core/src/main/java/org/hibernate/event/service/internal/EventListenerGroupImpl.java [hunk 2]
```diff
@@ -222,6 +222,7 @@ private boolean listenerShouldGetAdded(T listener) {
 							checkAgainstBaseInterface( listener );
 							performInjections( listener );
 							itr.set( listener );
+							listeners = null; //Marks it for refreshing
```
Comment (dreab8) [inline hibernate-core/src/main/java/org/hibernate/event/service/internal/EventListenerGroupImpl.java]: not 100% sure but I think the correct thing to do is to remove the original listener from `listeners` .

Comment (DavideD) [reply hibernate-core/src/main/java/org/hibernate/event/service/internal/EventListenerGroupImpl.java]: That does not seem the approach used in the class, though

**Comment by DavideD** [comment]
I'm going to create an issue

File: hibernate-core/src/main/java/org/hibernate/event/service/internal/EventListenerGroupImpl.java (hunk 2 above)
Comment (dreab8) [reply hibernate-core/src/main/java/org/hibernate/event/service/internal/EventListenerGroupImpl.java]: forget my comment, I saw how `listerner` it is used

**Comment by DavideD** [comment]
By the way, the end effect is that some listeners cannot be replaced. For example: `DefaultInitializeCollectionEventListener`

**Comment by sebersole** [comment]
The clarify my comment... the real problem in the code block you highlight is both fields are not updated.  Only `#listenersAsList` is updated.  But its `#listeners` that is iterated to handle the events.  So that code block gets the 2 out of sync

File: hibernate-core/src/test/java/org/hibernate/event/service/internal/EventListenerReplacementStrategyTest.java (hunk 1 above)
Comment (DavideD) [reply hibernate-core/src/test/java/org/hibernate/event/service/internal/EventListenerReplacementStrategyTest.java]: > I'm a little nervous about this. Why not just check for each being an `instanceof ClearEventListener`? The intention of your test is clearly that there be only one.

OK, I will change it

> I'd also consider asserting `org.hibernate.event.service.spi.EventListenerGroup#listeners` to make sure that the listener actually got replaced/appended. Coupled with the point above, its possible that your code itself is not actually appending/replacing the listener

OK,  will also test listeners()

**Comment by DavideD** [comment]
> But its `#listeners` that is iterated to handle the events. So that code block gets the 2 out of sync

Yes, I was surprised we didn't notice this before though.
It seems that the code uses the iterator from `#listenerAsList` more often than the listeners array.

**Comment by DavideD** [comment]
I will close this PR and have a look to see what happens if I only keep the array field.

**Pull request closed without merging**